rest/route/pod_agent.go

```diff
@@ -115,13 +121,91 @@ func (h *podAgentNextTask) Parse(ctx context.Context, r *http.Request) error {
 }
 
 func (h *podAgentNextTask) Run(ctx context.Context) gimlet.Responder {
-	j := units.NewPodTerminationJob(h.podID, "reached end of pod lifecycle", utility.RoundPartOfMinute(0))
-	if err := amboy.EnqueueUniqueJob(ctx, h.env.RemoteQueue(), j); err != nil {
+	p, err := h.findPod()
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+
+	h.setAgentFirstContactTime(p)
+
+	pd, err := h.findDispatcher()
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+
+	nextTask, err := pd.AssignNextTask(ctx, h.env, p)
+	if err != nil {
 		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
 			StatusCode: http.StatusInternalServerError,
-			Message:    errors.Wrap(err, "enqueueing job to terminate pod").Error(),
+			Message:    errors.Wrap(err, "dispatching next task").Error(),
 		})
 	}
+	if nextTask == nil {
+		j := units.NewPodTerminationJob(h.podID, "reached end of pod lifecycle", utility.RoundPartOfMinute(0))
+		if err := amboy.EnqueueUniqueJob(ctx, h.env.RemoteQueue(), j); err != nil {
+			return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+				StatusCode: http.StatusInternalServerError,
+				Message:    errors.Wrap(err, "enqueueing job to terminate pod").Error(),
+			})
+		}
+	}
 
 	return gimlet.NewJSONResponse(apimodels.NextTaskResponse{})
 }
+
+func (h *podAgentNextTask) findPod() (*pod.Pod, error) {
+	p, err := pod.FindOneByID(h.podID)
+	if err != nil {
+		return nil, gimlet.ErrorResponse{
+			StatusCode: http.StatusInternalServerError,
+			Message:    errors.Wrap(err, "finding pod").Error(),
+		}
+	}
+	if p == nil {
+		return nil, gimlet.ErrorResponse{
+			StatusCode: http.StatusNotFound,
+			Message:    fmt.Sprintf("pod '%s' not found", h.podID),
+		}
+	}
+
+	return p, nil
+}
+
+func (h *podAgentNextTask) setAgentFirstContactTime(p *pod.Pod) {
+	if !p.TimeInfo.Initializing.IsZero() {
```

Comment:
.Initializing was a slightly confusing name to me at first. Can we renamed to something like InitializedTime (to keep consistent with past tense timestamp names like LastCommunicated).

Reply:
It's already in a `TimeInfo` struct whose sole purpose is to hold timing information about when certain things happened to the pod, so it seems pretty clear. I don't generally use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation) since it's redundant to the `TimeInfo` struct and modern IDEs include type information that it's a `time.Time` anyways.

Reply:
Ok, I'm fine with keeping it as is.